### ISSUE TITLE
compute_pixel_region_variance reads outside image boundaries

### ISSUE THREAD

**Issue by alecazam** [ARM-software/astc-encoder#164]
I was adding float4 support and this routine kept reading outside the array.  It seems to use +/radius without bothering to clamp the index each of z_src/y_src/x_src before the read.  This is my working version of that, but you can apply the same to the 8/16F cases. 

```

static void compute_pixel_region_variance(
	astcenc_context& ctx,
	const pixel_region_variance_args* arg
) {

if (img->data_type == ASTCENC_TYPE_F32)
    {
        float4* data32f = static_cast<float4*>(img->data);

// USE_2D_ARRAY
        
        // Swizzle data structure 4 = ZERO, 5 = ONE (in FP16)
//        uint16_t data[6];
//        data[ASTCENC_SWZ_0] = 0;
//        data[ASTCENC_SWZ_1] = 0x3C00;

        for (int z = zd_start; z < padsize_z; z++)
        {
            int z_src = (z - zd_start) + src_offset_z - kernel_radius_z;

            if (z_src < 0)
                z_src = 0;
            if (z_src > img->dim_z)
                z_src = img->dim_z-1;
            
            for (int y = 1; y < padsize_y; y++)
            {
                int y_src = (y - 1) + src_offset_y - kernel_radius_xy;
                
                if (y_src < 0)
                    y_src = 0;
                if (y_src > img->dim_y)
                    y_src = img->dim_y-1;
                
                for (int x = 1; x < padsize_x; x++)
                {
                    int x_src = (x - 1) + src_offset_x - kernel_radius_xy;
                    
                    /* Note: not using swizzling
                    data[0] = data16[z_src][y_src][4 * x_src];
                    data[1] = data16[z_src][y_src][4 * x_src + 1];
                    data[2] = data16[z_src][y_src][4 * x_src + 2];
                    data[3] = data16[z_src][y_src][4 * x_src + 3];

                    uint16_t r = data[swz.r];
                    uint16_t g = data[swz.g];
                    uint16_t b = data[swz.b];
                    uint16_t a = data[swz.a];

                    float4 d = float4(sf16_to_float(r),
                                      sf16_to_float(g),
                                      sf16_to_float(b),
                                      sf16_to_float(a));
                     */
                    
                    // this call just reads right off the array, had to add these clamps
                    if (x_src < 0)
                        x_src = 0;
                    if (x_src > img->dim_x)
                        x_src = img->dim_x-1;
                    
                    float4 d = data32f[y_src * img->dim_x + x_src];

                    if (!are_powers_1)
                    {
                        d.x = powf(MAX(d.x, 1e-6f), rgb_power);
                        d.y = powf(MAX(d.y, 1e-6f), rgb_power);
                        d.z = powf(MAX(d.z, 1e-6f), rgb_power);
                        d.w = powf(MAX(d.w, 1e-6f), alpha_power);
                    }

                    VARBUF1(z, y, x) = d;
                    VARBUF2(z, y, x) = d * d;
                }
            }
        }
    }

```

**Comment by solidpixel** [comment]
In the current code the user must pre-pad the input data, replicating the edge texels, if they are using any region-based error heuristic. 

https://github.com/ARM-software/astc-encoder/blob/ab16aaabe4402361a12c7553c817519ed65fc31e/Source/astcenc.h#L95

Not needing this would allow zero copy inputs, so would be a good idea ...

**Comment by solidpixel** [comment]
Fixed for 2.2 via a288266.
